### PR TITLE
Configurable Node Operational Timeouts

### LINKED ISSUE #3155: Improve Handling of Timeouts
Timeouts could be better configured:

- An additional Porter request parameter to specify timeouts
- Python API allows timeout to be specified as part of PRERetrievalClient. Ideally the PRERetrievalClient would use concurrency (i.e. the WorkerPool which it currently does not) and an overall timeout used. The ThresholdDecryptionClient does this for the cbd /decrypt operation.

A stopgap measure was introduced in https://github.com/nucypher/nucypher/pull/3153 by @derekpierre 

We would like a better solution for v8.0.0.

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:** 
- [ ] 1
- [ ] 2
- [X] 3

**What this does:**
Allows timeouts to be programmatically specified for TACo operations when connecting to nodes i.e. `threshold_decrypt` / `retrieve_and_decrypt` (PRE).

It would allow the ability for Porter to accept timeout parameters in requests that go through Porter - so then apps can potentially use their own timeout (note: for apps to do that `nucypher-ts` (`taco-api`) may need to be modified.

**Issues fixed/closed:**
> - Fixes #...

Related to:
- #3153 
- https://github.com/nucypher/nucypher-porter/issues/49

Fixes #3155 .

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

### CODE REVIEW

File: nucypher/characters/lawful.py [hunk 1]
```diff
@@ -510,6 +510,7 @@ def retrieve(
         alice_verifying_key: PublicKey,  # KeyFrag signer's key
         encrypted_treasure_map: EncryptedTreasureMap,
         publisher_verifying_key: Optional[PublicKey] = None,
+        timeout: int = 10,
```
Comment (cygnusv) [inline nucypher/characters/lawful.py:513]: Can we make the `10` as a constant value somewhere?

File: nucypher/characters/lawful.py [hunk 2]
```diff
@@ -723,19 +707,14 @@ def threshold_decrypt(
         threshold_message_kit: ThresholdMessageKit,
         context: Optional[dict] = None,
         ursulas: Optional[List["Ursula"]] = None,
-        peering_timeout: int = 60,
+        decryption_timeout: int = 15,
```
Comment (cygnusv) [inline nucypher/characters/lawful.py:710]: same as above, can we make this a constant value?

File: nucypher/network/decryption.py [hunk 3]
```diff
@@ -27,15 +27,15 @@ def gather_encrypted_decryption_shares(
         self,
         encrypted_requests: Dict[ChecksumAddress, EncryptedThresholdDecryptionRequest],
         threshold: int,
-        timeout: float = 15,
+        timeout: int = 15,
```
Comment (cygnusv) [inline nucypher/network/decryption.py:30]: Same comment

File: nucypher/network/middleware.py [hunk 4]
```diff
@@ -267,7 +293,7 @@ def reencrypt(
         self,
         ursula: "characters.lawful.Ursula",
         reencryption_request_bytes: bytes,
-        timeout=8,
+        timeout: int = 10,
```
Comment (cygnusv) [inline nucypher/network/middleware.py:296]: Same comment

File: nucypher/network/middleware.py [hunk 5]
```diff
@@ -281,7 +307,7 @@ def get_encrypted_decryption_share(
         self,
         ursula: "characters.lawful.Ursula",
         decryption_request_bytes: bytes,
-        timeout=8,
+        timeout: int = 15,
```
Comment (cygnusv) [inline nucypher/network/middleware.py:310]: Same

File: nucypher/network/retrieval.py [hunk 6]
```diff
@@ -250,14 +253,16 @@ def _request_reencryption(
...
-                        **context) -> Tuple[List[RetrievalResult], List[RetrievalError]]:
-
+    def retrieve_cfrags(
+        self,
+        treasure_map: TreasureMap,
+        retrieval_kits: Sequence[RetrievalKit],
+        alice_verifying_key: PublicKey,  # KeyFrag signer's key
+        bob_encrypting_key: PublicKey,  # User's public key (reencryption target)
+        bob_verifying_key: PublicKey,
+        timeout: int = 10,
```
Comment (cygnusv) [inline nucypher/network/retrieval.py:263]: Same

**Review by cygnusv — COMMENTED** [review]
Looks good, my major comment is to extract the constant values for tiemouts to constants. That can be in their respective class, module, or in a constants module.

File: nucypher/characters/lawful.py (hunk 1 above)
Comment (derekpierre) [reply nucypher/characters/lawful.py]: ✔️

File: nucypher/network/decryption.py (hunk 3 above)
Comment (derekpierre) [reply nucypher/network/decryption.py]: ✔️

File: nucypher/network/middleware.py (hunk 5 above)
Comment (derekpierre) [reply nucypher/network/middleware.py]: ✔️

File: nucypher/network/middleware.py (hunk 4 above)
Comment (derekpierre) [reply nucypher/network/middleware.py]: ✔️

File: nucypher/network/retrieval.py (hunk 6 above)
Comment (derekpierre) [reply nucypher/network/retrieval.py]: ✔️

File: nucypher/characters/lawful.py (hunk 2 above)
Comment (derekpierre) [reply nucypher/characters/lawful.py]: ✔️

File: nucypher/network/middleware.py [hunk 7]
```diff
@@ -21,10 +22,21 @@
 SSL_LOGGER = Logger('ssl-middleware')
 EXEMPT_FROM_VERIFICATION.bool_value(False)
 
+# It’s a good practice to set connect timeouts to slightly larger
+# than a multiple of 3, which is the default TCP packet retransmission window.
+# (https://requests.readthedocs.io/en/latest/user/advanced/#timeouts)
```
Comment (piotr-roslaniec) [inline nucypher/network/middleware.py:27]: Very cool

File: nucypher/network/middleware.py [hunk 8]
```diff
@@ -102,11 +115,22 @@ def parse_node_or_host_and_port(self, node, host, port):
             raise ValueError("You need to pass either the node or a host and port.")
         return host, port, self.library
 
+    def _determine_timeout(self, provided_timeout) -> Union[float, Tuple[float, float]]:
```
Comment (piotr-roslaniec) [inline nucypher/network/middleware.py:118]:
```suggestion
    def _determine_timeout(self, provided_timeout=None) -> Union[float, Tuple[float, float]]:
```

File: nucypher/network/middleware.py [hunk 9]
```diff
@@ -102,11 +115,22 @@ def parse_node_or_host_and_port(self, node, host, port):
...
+        # provided timeout (could be None) is really for reading not initial connection timeout,
+        # but use as fallback for connect timeout
```
Comment (piotr-roslaniec) [inline nucypher/network/middleware.py:120]: I didn't get that. How are these two used? Is `connect_timeout` for the TCP connection or finding peers? Is `read_timeout` for uhh reading the data from said request?

**Review by piotr-roslaniec — APPROVED** [review]

File: nucypher/network/middleware.py (hunk 9 above)
Comment (derekpierre) [reply nucypher/network/middleware.py]: Yes. Basically there are two timeouts for the `requests` library:
- `connect timeout`: number of seconds Requests will wait for your client to establish a connection to a remote machine
- `read timeout`: number of seconds the client will wait for the server to send a response

If you only specify one value - then the same value is used for both. When someone provides a timeout via the method call, they are really intending to specify the `read timeout`, so we use our own internal connect timeout of 3.05 seconds, which is sufficient for connection to nodes. 

The read timeout can be unnecessarily large if we use it for the connect timeout.

Modified the code a little to attempt to make that clearer.

(see https://requests.readthedocs.io/en/latest/user/advanced/#timeouts)

**Pull request merged**